File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        substrMatch = False
        short = str1 if len(str1) <= len(str2) else str2
        long = str2 if len(str1) <= len(str2) else str1
        originalShort = short

        while not substrMatch:

            if short == "": substrMatch = True 
        
            elif (short == long[:len(short)] and len(long) % len(short) + 
                                                (len(str1) % len(short)) +
                                                (len(str2) % len(short)) == 0):
                substrMatch = True
                for i in range(0, len(originalShort), len(short)):
                    if short != originalShort[i:i+len(short)]:
                        substrMatch = False
                        short = short[:-1]
                        break
                if substrMatch:
                    for i in range(0, len(long), len(short)):
                        if short != long[i:i+len(short)]:
                            substrMatch = False
                            short = short[:-1]
                            break

            else:
                short = short[:-1]

        return short

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.cpp
static const bool a() {
    std::ios_base::sync_with_stdio(false);
    std::cout.tie(nullptr);
    std::cin.tie(nullptr);
    return true;
}
class Solution {
public:
    string gcdOfStrings(const string str1, const string str2) {
        const string* longStr = str1.length() >= str2.length() ? &str1 : &str2;
        const string* shortStr = str1.length() < str2.length() ? &str1 : &str2;
        string str = *shortStr;
        while (str.length() > 0) {
            cout << str << endl;
            const int sizeL = longStr->length();
            const int sizeS = shortStr->length();
            const int sizeC = str.length();
            if (sizeL % sizeC == 0 && sizeS % sizeC == 0) {
                if (checkSubstr(longStr, shortStr, str)) {
                    return str;
                }
            }
            str.pop_back();
        }
        return "";
    }

    bool checkSubstr(const string* longStr, const string* shortStr, const string strToCheck) {
        const int sizeL = longStr->length();
        const int sizeS = shortStr->length();
        const int sizeC = strToCheck.length();
        for (int i = 0; i < sizeL; i += sizeC) {
            const bool validS = 
                i + sizeC > sizeS || 
                strToCheck == shortStr->substr(i, sizeC) 
            ? true : false;
            const bool validL = strToCheck == longStr->substr(i, sizeC);
            if (!validS || !validL) return false;
        }
        return true;
    }
};